src/main/java/jenkins/branch/MultiBranchProject.java

```diff
@@ -2020,253 +2178,85 @@ public void observe(@NonNull SCMHead head, @NonNull SCMRevision revision) throws
                     printStackTrace(e, listener.error("Could not update last seen revision hash"));
                 }
-            } finally {
-                observer.completed(encodedName);
+            }
+            return scmLastSeenRevision;
+        }
+
+        private void doAutomaticBuilds(@NonNull SCMHead head, @NonNull SCMRevision revision, @NonNull String rawName, @NonNull P project, Action[] revisionActions, SCMRevision scmLastBuiltRevision, SCMRevision scmLastSeenRevision) {
+            if (isAutomaticBuild(head, revision, scmLastBuiltRevision, scmLastSeenRevision)) {
+                scheduleBuild(
+                        _factory,
+                        project,
+                        revision,
+                        listener,
+                        rawName,
+                        causeFactory.create(source),
+                        revisionActions
+                );
+            } else {
+                listener.getLogger().format("No automatic build triggered for %s%n", rawName);
+            }
+            try {
+                _factory.setLastSeenRevisionHash(project, revision);
+            } catch (IOException e) {
+                printStackTrace(e, listener.error("Could not update last seen revision hash"));
             }
         }
-    }
 
-    /**
-     * Tests if the specified {@link SCMHead} should be automatically built when discovered / modified.
-     * @param source the source.
-     * @param head the head.
-     * @param currRevision the current built revision.
-     * @param lastBuiltRevision the previous built revision
-     * @param listener the {@link TaskListener}
-     * @param lastSeenRevision the last seen revision
-     * @return {@code true} if the head should be automatically built when discovered / modified.
-     */
-    private boolean isAutomaticBuild(@NonNull SCMSource source,
-                                     @NonNull SCMHead head,
-                                     @NonNull SCMRevision currRevision,
-                                     @CheckForNull SCMRevision lastBuiltRevision,
-                                     @CheckForNull SCMRevision lastSeenRevision,
-                                     @NonNull TaskListener listener) {
-        BranchSource branchSource = null;
-        for (BranchSource s: sources) {
-            if (s.getSource().getId().equals(source.getId())) {
-                branchSource = s;
-                break;
+        /**
+         * Tests if the specified {@link SCMHead} should be automatically built when discovered / modified.
+         * @param head the head.
+         * @param currRevision the current built revision.
+         * @param lastBuiltRevision the previous built revision
+         * @param lastSeenRevision the last seen revision
+         * @return {@code true} if the head should be automatically built when discovered / modified.
+         */
+        private boolean isAutomaticBuild(@NonNull SCMHead head,
```

Comment:
Moved this method into the observer class.  There is not advantage to it being on the project instance.